### PR TITLE
Add method for tapering gate operations

### PR DESCRIPTION
**Pull request by obliviateandsurrender**
**Context:**
This PR adds the final missing piece 🧩 for the tapered variational routine.

**Description of the Change:**
Adds a function for tapering the gate operations, which are then exponentiated using `qml.PauliRot`. 

**Benefits:**
One can now taper individual gate operations and build tapered variants of templates. 

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

### CODE REVIEW

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:543]: I recommend checking with the operator arithmetic team to understand how they are thinking of checking commutativity between operators. We don't want to have duplicate functionality in the codebase

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:568]:
```suggestion
    tapered circuit ansatz.
```
one ansatz, two ansatze

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:574]:
```suggestion
        singles (list(list(int))): list with the indices `r`, `p` of the two qubits representing the single excitation
```

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:575]:
```suggestion
        doubles (list(list(int))): list with the indices `s`, `r`, `q`, `p` of the four qubits representing the double excitation
```

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:582]: I recommend changing the example since He doesn't really form molecules

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:584]: I forget what our usual practice is, but it may be good to use just `qchem` instead of `qml.qchem`

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:560]: This function is specifically tailored for tapering both single and double excitation operators simultaneously, which means users can't really leverage it to taper other gates.

Couldn't we use the same strategy as in here to taper an arbitrary gate? I think that would be very powerful

**Review by ixfoduap — CHANGES_REQUESTED** [review]

File: pennylane/qchem/tapering.py
Comment (josh146) [reply pennylane/qchem/tapering.py:543]: @PennyLaneAI/core @albi3ro

File: pennylane/qchem/tapering.py
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:560]: Yep, this strategy can taper any arbitrary gate for which we have the generators.

**Review by soranjh — CHANGES_REQUESTED** [review]

File: pennylane/qchem/tapering.py
Comment (albi3ro) [reply pennylane/qchem/tapering.py:543]: So we have a `qml.is_commuting` function, currently located in `ops/functions`.

The function isn't particularly general or easy to understand, but I do think we should merge both this logic and `qml.grouping.is_commuting` into that function.

Especially merging in ` qml.grouping.is_commuting`.  Rather confusing having two functions with the same names but different behaviours and locations in the code base.

Comment (albi3ro) [reply pennylane/qchem/tapering.py:543]: See Issue #3007 .  Core squad will try to combine `qml.grouping.is_commuting` and `qml.is_commuting` before the next release.

**Comment by obliviateandsurrender** [comment]
Based on the feedback, I have updated the previous `taper_excitation` method to a more general `taper_operation` method which can be used on any arbitrary gate.

File: tests/qchem/test_tapering.py
Comment (obliviateandsurrender) [reply tests/qchem/test_tapering.py]: I have removed this in the new tests.

File: tests/qchem/test_tapering.py
Comment (Jaybsoni) [inline tests/qchem/test_tapering.py:663]: One suggestion here, instead of doing a full workflow test we could just parameterize over the inputs of the taper_operation function for each case. This way we don't have to compute those values each time. 

If you think developers who see the test might want to know where it came from, then we can add comments beside each parameterization saying something like `# these ops come from the H2 molecule with ____ geometry and ____ charge`

This reduces the complexity of the test and focuses on specifically what you want to test.

**Review by Jaybsoni — CHANGES_REQUESTED** [review]
Hey, left some comments, some suggestions and some small fixes.

File: pennylane/qchem/tapering.py
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:620]: This validation check is for the generator operators given by the users in case PL doesn't implement one.

File: pennylane/qchem/tapering.py
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py]: Updated this with `ValueError`. In general, I wanted to raise my own custom exception message instead of the one raised by the `qml.generator`.

File: tests/qchem/test_tapering.py
Comment (obliviateandsurrender) [reply tests/qchem/test_tapering.py]: Great idea! I have updated these with their corresponding tapering data.

File: tests/qchem/test_tapering.py
Comment (obliviateandsurrender) [reply tests/qchem/test_tapering.py]: I have tried to reduce it without using the helper function for now. Let me know if that is still required.

File: pennylane/qchem/tapering.py
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:564]: Added to the docstring!

File: pennylane/qchem/tapering.py
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:552]: While I explain the working of the operation in docstring here, we were planning to include the theory/equation in the demo we have for the tapering. I already have an open [PR](https://github.com/PennyLaneAI/qml/pull/560) for that.

File: pennylane/qchem/tapering.py
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:616]: Quick note: Once this is fixed and `qml.equal` supports `qml.exp` then we can use it instead of `qml.PauliRot` in the final step as well.

File: pennylane/qchem/tapering.py
Comment (Jaybsoni) [reply pennylane/qchem/tapering.py:552]: Sounds good, then in the future I also recommend adding a link in these docs to that demo, so that there is a link to an explanation in the docs.

File: pennylane/qchem/tapering.py
Comment (Jaybsoni) [reply pennylane/qchem/tapering.py:616]: Sounds good!

File: tests/qchem/test_tapering.py
Comment (Jaybsoni) [inline tests/qchem/test_tapering.py:724]: This looks like it can be further simplified by wrapping the logic in a for loop and iterating over `[singles, doubles ]`

File: tests/qchem/test_tapering.py
Comment (Jaybsoni) [inline tests/qchem/test_tapering.py:821]: Just a suggestion, feel free to ignore. We could probably also remove this and have the true results stored directly (with a comment mentioning how the tapered operation was computed).

File: tests/qchem/test_tapering.py
Comment (Jaybsoni) [reply tests/qchem/test_tapering.py]: It still looks big and "scary", but also we don't need to worry to much about how elegant our test suite looks.

**Review by Jaybsoni — COMMENTED** [review]
Mostly looking good on my side, just a few minor comments to clean up the tests. I am not too familiar with the theory of tapering so maybe some like @soranjh can comment on  theory aspect of the implementation. 

Happy to approve once the comments are addressed

File: pennylane/qchem/tapering.py
Comment (soranjh) [inline pennylane/qchem/tapering.py:554]: For consistency with the taper function, I suggest these changes:

generators (list[Hamiltonian]): generators expressed as PennyLane Hamiltonians
paulix_sector (llist[int]): eigenvalues of the Pauli-X operators

File: pennylane/qchem/tapering.py
Comment (soranjh) [inline pennylane/qchem/tapering.py:563]: unitary operation"s"?

File: pennylane/qchem/tapering.py
Comment (soranjh) [inline pennylane/qchem/tapering.py:567]: The operation needs to be the generator of the operation?

File: pennylane/qchem/tapering.py
Comment (soranjh) [reply pennylane/qchem/tapering.py:563]: Why it has to be a PauliRot type?

File: tests/qchem/test_tapering.py
Comment (soranjh) [inline tests/qchem/test_tapering.py:633]: Might consider a test for computing the ground state energy with the tapered excitation gates for a simple molecule.

File: pennylane/qchem/tapering.py
Comment (soranjh) [inline pennylane/qchem/tapering.py:593]: I suggest adding a theory section and explain the details of how the operation is tapered, using:

    .. details::
        :title: Theory

        Explain the theory ...

**Review by soranjh — CHANGES_REQUESTED** [review]

File: pennylane/qchem/tapering.py
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:563]: It should be `unitary operation` as we are tapering one operation at a time. They have to be of type `PauliRot` since we are using that to exponentiate the tapered generator operator.

File: pennylane/qchem/tapering.py
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:567]: Actually, the argument `gen_op` needs to be the generator of the operation. I will make it more explicit here.

File: pennylane/qchem/tapering.py
Comment (timmysilv) [inline pennylane/qchem/tapering.py:629]:
```suggestion
                qml.is_commuting(op1, op2)
```
(once latest master is merged here now that #3033 was merged)

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:577]: This Comment applies also to the example code in the changelog: why is the gate parameter `1`? I find that a confusing choice since I'd expect a float

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:581]:
```suggestion
    This can even be used within a :class:`~.pennylane.QNode`:
```

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:17]: This is a general function that can in principle be used for any operation. I do not think it should be inside qchem, but rather be more visibly to general users of pennylane. The same applies to basically all tapering functionality, except I believe `taper_hf`. Let's consider doing this at some point

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:604]:
```suggestion
        :math:`\tau`, we should have :math:`[V, \tau_i] = 0\ \forall \theta \in \mathbb{R},\ \forall \tau_i`.
```

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:605]:
```suggestion
        This would hold only when its generator itself commutes with each :math:`\tau_i`,
```
Please remember to always check the built docs before opening a PR for review

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:604]: This is a bit too formal for my taste. Consider removing `\in \mathbb{R}` and use words "for all" rather than the symbol `\forall`. Same elsewhere in the example

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:619]: I would remove this last sentence as it is obvious that parametrized gates are strictly more general

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:595]: Super nice theory section!

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:643]:
```suggestion
                f"Generator for the operation needs to be a qml.Hamiltonian, but got {type(gen_op)}."
```

Comment (ixfoduap) [reply pennylane/qchem/tapering.py:643]: Attention to details!

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [inline pennylane/qchem/tapering.py:647]: I wonder if there is a smarter way to obtain the phase, and whether it matters. If the generator matrix is obtained from `qml.generator()`, then shouldn't `expm(1j * matrix)` give back the right matrix for the unitary describing the operation?

Comment (ixfoduap) [reply pennylane/qchem/tapering.py:647]: Oh I see, this is in case the user provider their own? Maybe consider adding a short comment to the code, e.g.,
`# check that user-provided generator is correct`

File: tests/qchem/test_tapering.py
Comment (ixfoduap) [inline tests/qchem/test_tapering.py:847]: I'm not sure this integration test makes sense to include here. It's good to make these checks during development, but it's problematic to include such expensive tests in the code. Please remove this

**Review by ixfoduap — CHANGES_REQUESTED** [review]

File: pennylane/qchem/tapering.py
Comment (soranjh) [inline pennylane/qchem/tapering.py:604]:
```suggestion
        :math:`\tau`, we should have :math:`[V, \tau_i] = 0\ \forall \theta \in \mathbb{R},\ \forall \tau_i`.
```

File: pennylane/qchem/tapering.py
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:577]: This was rather arbitrary. I am now updating it to `np.pi`.

File: pennylane/qchem/tapering.py
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py]: We have a similar docstring for the other taper functions as well. Should we consider changing all of them @soranjh ?

File: tests/qchem/test_tapering.py
Comment (obliviateandsurrender) [reply tests/qchem/test_tapering.py]: I have modified the tests to check the consistencies in the expectation values of tapered observables with respect to tapered states evolved with the tapered operation. The observables include hamiltonian and arbitrary Pauli sum, and operations are both parameterized and unparameterized gates.

File: pennylane/qchem/tapering.py
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:565]: I have seen it being used in a lot of places in the codebase beyond `qchem`. @Jaybsoni can confirm as well.

File: pennylane/qchem/tapering.py
Comment (soranjh) [reply pennylane/qchem/tapering.py:17]: The `taper` function is accessible as `qml.taper`, we can do the same for `taper_operation` if it is not qchem-specific anymore.

File: pennylane/qchem/tapering.py
Comment (soranjh) [reply pennylane/qchem/tapering.py]: Maybe we can change them to: "Transform a gate operation with a Clifford operator and then taper qubits."?

**Review by soranjh — APPROVED** [review]

File: pennylane/qchem/tapering.py
Comment (ixfoduap) [reply pennylane/qchem/tapering.py:17]: Yes, let's do that. As far as I can tell, this method in principle applies to any operation

**Review by ixfoduap — APPROVED** [review]

File: pennylane/qchem/tapering.py
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:17]: Added `taper_operation` to __init__.

**Pull request merged**
